ansible/testbed-cli.sh

```diff
@@ -644,6 +652,103 @@ function cleanup_vmhost
+  for topology in $topologies_by_setup_name
+  do
+      if [[ "$topology" == *"$setup_name"* ]]; then
+          result_topologies_list+=("$topology")
+      fi
+  done
+
+  echo ${result_topologies_list[@]}
+}
+
+function is_cache_exist
+{
+  cache_files_path="$(pwd)/cached_topologies_path"
+  if [ ! -f "$cache_files_path" ]; then
+      echo "cached_topologies_path file does not exist, the path for cached topologies is mandatory to run this command. exiting..." >&2
+      echo ""
+      return
+  fi
+
+  cache_files_path_value=$(cat $cache_files_path)
+  if [[ "$cache_files_path_value" == "" ]]; then
+      echo "cached_topologies_path file content is empty, please add a path to cache topologies. exiting..." >&2
+      echo ""
+      return
+  fi
+
+  if [ ! -d "$cache_files_path_value" ]; then
+      echo "Path specified in cached_topologies_path file does not exist, creating..." >&2
+      mkdir -p "$cache_files_path_value"
+  fi
+
+  echo $cache_files_path_value
+}
+
+function deploy_topo_with_cache
+{
+  testbed_name=$1
+  inventory=$2
+  passwd=$3
+
+  cache_files_path_value=$(is_cache_exist)
+  if [[ -z $cache_files_path_value ]]; then
+    exit
+  fi
+
+  read_file ${testbed_name}
+  setup_name=$dut
+  if [[ "$setup_name" == "" ]]; then
+      echo "No such testbed: $testbed_name, exiting..."
+      exit
+  fi
+  setup_topologies=$(read_topologies_from_csv_file)
+
+  if [[ ! " ${setup_topologies[*]} " =~ " ${testbed_name} " ]]; then
+      echo "No such testbed: $testbed_name, exiting..."
+      exit
+  fi
+
+  remove_all_topologies=true
+
+  echo "Try to read the topology from cache file: '$cache_files_path_value/$setup_name'"
+  if [ -f "$cache_files_path_value/$setup_name" ]; then
+      echo "Cache file: '$setup_name', exists."
+      cache_topo=$(cat $cache_files_path_value/$setup_name)
+      if [[ ! " ${setup_topologies[*]} " =~ " ${cache_topo} " ]]; then
+          echo "Topology in cache file [$setup_name] is not valid, falling back."
+          echo "Removing all known topologies from the setup and creating new cache file"
+      elif [[ "$cache_topo" == "$testbed_name" ]]; then
+          echo "Topology [$testbed_name] is already deployed, exiting..."
+          exit
+      else
+          echo "Removing cached topology [$cache_topo] from the setup and updating the cache file"
+          remove_all_topologies=false
+      fi
+  else
+      echo "Cache file: '$cache_files_path_value/$setup_name', not exists."
+      echo "Removing all known topologies from the setup and creating new cache file."
+  fi
+
+  if [ "$remove_all_topologies" = true ]; then
+      for topo in $setup_topologies
+      do
+          ./testbed-cli.sh remove-topo $topo $passwd
+      done
+  else
+      .testbed-cli.sh remove-topo $cache_topo $passwd
+  fi
+
+  ./testbed-cli.sh add-topo $testbed_name $passwd
+  ./testbed-cli.sh deploy-mg $testbed_name $inventory $passwd
```

Comment:
Why call the script itself here? Why not call function `add_topo` and `deploy_minigraph ` instead?

Reply:
@liorghub Can you elaborate this?

Reply:
No need to call the script itself, I have changed it to function calls.